V1.1.9

devhawala opened the pull request:
This pull-request contains several improvements and changes to Darkstar 1.1.0.0 collected over time.
- The commit message to the first commit in branch v1.1.9 contains a list of the code changes, most important are improvements to the network interface and the ability to boot the guest system from the command line through the configuration file, both on Windows and Unixoids.
- The readme files were updated accordingly.
- The version number 1.1.9.0 was chosen only to keep the version in the forked repository out of the way from possibly new versions in the original repo. So feel free to use a differing version matching the main versioning sequence for Darkstar.
- Finally a disk image for StarOS 5.0 was added, with a matching configuration file; this is also an example for a Darkstar configuration file having all parameters with some comments.

jdersch commented on D/Display/DisplayController.cs:
```diff
@@ -171,7 +171,7 @@ private void HorizontalRetraceCallback(ulong skewNsec, object context)
                     // Normal line : 32 bits of border pattern, 1024 bits of display, 32 bits of border pattern
                     // Border pattern: low byte on lines 4n, 4n+1; high byte on 4n+2, 4n+3.
-                    int patternByte = (effectiveScanline & 0x2) == 0 ? _displayBorder & 0xff : _displayBorder >> 8;
+                    int patternByte = (effectiveScanline & 0x2) == 2 ? _displayBorder & 0xff : _displayBorder >> 8;
```
The bug here is more subtle -- changing this here provides the correct pattern alignment for Viewpoint, but breaks the pattern on Interlisp-D.  (Currently Interlisp-D looks fine, VP is broken).  Prefer not changing this now until the root cause can be determined.

livingcomputermuseum replied:
The root cause of this issue is more subtle.  Without this change, Interlisp-D's border pattern aligns properly and Viewpoint is misaligned; with this change the opposite will be true.  I suspect the root cause is a timing issue during the start of the frame that I have to dig into.  I'd prefer not to change this right now.  (See also line 216.)

livingcomputermuseum commented on D/Darkstar.csproj:
```diff
@@ -37,26 +45,40 @@
   <PropertyGroup>
     <ApplicationIcon>Dandelion.ico</ApplicationIcon>
   </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)' == 'Release' ">
+    <CheckForOverflowUnderflow>False</CheckForOverflowUnderflow>
+    <BaseIntermediateOutputPath>obj\</BaseIntermediateOutputPath>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Platform)' == 'AnyCPU' ">
+    <BaseAddress>4194304</BaseAddress>
+    <Prefer32Bit>True</Prefer32Bit>
+    <RegisterForComInterop>False</RegisterForComInterop>
+    <GenerateSerializationAssemblies>Auto</GenerateSerializationAssemblies>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Platform)' == 'x64' ">
+    <Prefer32Bit>true</Prefer32Bit>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+    <StartAction>Project</StartAction>
+    <StartArguments>-config xde_5.0.cfg</StartArguments>
+    <StartWorkingDirectory>C:\_Hans\Darkstar_NewFeatures\D</StartWorkingDirectory>
+  </PropertyGroup>
```
Please remove these extraneous changes to the project

livingcomputermuseum replied:
Actually, are any of these changes necessary?

livingcomputermuseum commented on D/Ethernet/EthernetController.cs:
```diff
@@ -624,12 +624,14 @@ private void OnHostPacketReceived(MemoryStream data)
                 _readerLock.ExitWriteLock();
 
                 if (Log.Enabled) Log.Write(LogComponent.EthernetReceive, "Packet (length {0}) added to pending buffer.", data.Length);
+                //(debug) Console.Write("  -- Packet (length {0}) added to pending buffer -- currently enqueued: {1}\n", data.Length, _pendingPackets.Count);
             }
             else
             {
                 //
                 // Too many queued-up packets, drop this one.
                 //
+                //(debug) Console.WriteLine("Pending buffer full; dropping this packet.");
```
Please remove commented out debug lines.

livingcomputermuseum commented on D/Ethernet/EthernetController.cs:
```diff
@@ -720,6 +722,12 @@ private void ReceiverPollCallback(ulong skewNsec, object context)
                         _inputPacket.Enqueue((ushort)(packetStream.ReadByte() << 8));
                     }
 
+                    // add 2 words to ensure that the ethernet task microcode does not ignore the packet
+                    // (the microcode seems to expect a 4 byte "frame check sequence" at the packet end)
+                    _inputPacket.Enqueue((ushort)0);
+                    _inputPacket.Enqueue((ushort)0);
```
Are you certain the microcode expects this and that there isn't a different root cause of this issue?  The Ethernet code as it stands works fine under Interlisp-D (tested with the TCP/IP stack) and as far as I know the Ethernet microcode between Interlisp and VP is basically identical.

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+using System;
+using System.IO;
+using System.Net.Sockets;
+using System.Threading;
+
+namespace D.Ethernet
+{
+	/// <summary>
+	/// Packet interface for accessing a Dodo NetHub as network device for Darkstar.
+	/// </summary>
+	public class NethubInterface : IPacketInterface
+	{
+    	
+		// this is the name of the NetHub network-device in the configuration dialog resp. the configuration file
+    	public const string NETHUB_NAME = "[[ Dodo-Nethub ]]";
+    	
+		private TcpClient client = null;
+		private NetworkStream stream = null;
+		
+		private ulong localAddress;
+		private ulong broadcastAddress;
+		
+		private ReceivePacketDelegate receiver = null;
+		private Thread receiverThread = null;
+		
+		private Byte[] sendBuffer = new byte[1026]; // this assumes that there is only ONE processing thread (which is the case for DarkStar!) 
+		
+		public NethubInterface()
+		{
+            client = new TcpClient(Configuration.NetHubHost, Configuration.NetHubPort);
+			stream = client.GetStream();
+			
+			this.localAddress = Configuration.HostID;
```
Un-tabify this file, please (4 spaces per tab).

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        /// Registers a callback delegate to handle packets that are received.
+        /// </summary>
+        /// <param name="callback"></param>
+        void IPacketInterface.RegisterReceiveCallback(ReceivePacketDelegate callback) {
```
Fix brace style (here and elsewhere in this file.)

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        	}
+        	
+        	int byteLen = packet.Length * 2;
+            //(debug) Console.Write("** sending packed, len = {0} bytes\n", byteLen);
```
Please remove this line (or make it a Log.Write call.)

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        /// Shuts down the encapsulation provider.
+        /// </summary>
+        void IPacketInterface.Shutdown() {
+        	//(debug) Console.WriteLine("** Stopping nethub interface...");
```
Same as above.

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        }
+        
+        private int getLenPrefix(NetworkStream str) {
+        	Int32 b1 = this.getByte(str) & 0xFF;
+        	Int32 b2 = this.getByte(str) & 0xFF;
```
Prefer "int" rather than "Int32"

livingcomputermuseum commented on D/Program.cs:
```diff
@@ -77,6 +78,10 @@ static void Main(string[] args)
                             break;
 
+                        case "-start":
+                            doStart = true;
```
Can you just set Configuration.Start directly here and eliminate doStart?

livingcomputermuseum commented on D/UI/ConfigurationDialog.cs:
```diff
@@ -123,13 +125,19 @@ private void PopulateNetworkAdapterList()
             // Populate the list with the interfaces available on the machine, if any.
             //
-            EthernetInterfaceListBox.Enabled = Configuration.HostRawEthernetInterfacesAvailable;
+            EthernetInterfaceListBox.Enabled = true; // Configuration.HostRawEthernetInterfacesAvailable;
```
Just remove this line entirely.

livingcomputermuseum commented on D/UI/ConfigurationDialog.cs:
```diff
@@ -252,5 +262,22 @@ private void OnClosed(object sender, FormClosedEventArgs e)
+        {
+            if (String.IsNullOrEmpty(txtNetHubHost.Text))
+            {
+                MessageBox.Show("The NetHub host can not be empty");
```
"cannot"

livingcomputermuseum commented on D/UI/DWindow.cs:
```diff
@@ -101,6 +101,21 @@ public DWindow(DSystem system)
+        
+        private bool firstShown = true;
+        
+		protected override void OnShown(EventArgs e)
+		{
+			base.OnShown(e);
+			if (firstShown) {
+				firstShown = false;
+				if (Configuration.Start && !_system.IsExecuting)
+	            {
+	                SystemExecutionContext context = new SystemExecutionContext(null, null, null, OnExecutionError);
+	                _system.StartExecution(context);
+	            }
+			}
+		}
```
Fix tabs.

livingcomputermuseum commented on D/UI/DWindow.cs:
```diff
@@ -529,6 +550,7 @@ private void UpdateFloppyDriveLabel()
 
         private void PopulateAltBoot()
         {
+        	_system.IOP.MiscIO.AltBoot = Configuration.AltBootMode;
```
Fix tabs.

livingcomputermuseum commented on D/readme.txt:
```diff
@@ -57,7 +57,9 @@ use.
 If you wish to make use of the emulated Star's Ethernet interface, you
 will need to have WinPCAP installed (on Windows) or libpcap (on the Unix of 
-your choice).  WinPCAP can be downloaded from http://www.winpcap.org/.
+your choice).  WinPCAP can be downloaded from http://www.winpcap.org/.    
+If only Dodo emulated XNS services are to be accessed from Darkstar, a direct
+connection to the Dodo NetHub can be configured instead of using a real network. 
```
Suggest adding a section somewhere covering Dodo (what it is, where to get it, etc...)

livingcomputermuseum commented on D/Ethernet/EthernetController.cs:
```diff
@@ -848,9 +856,12 @@ private void AttachHostEthernet()
+            	if (Configuration.HostPacketInterfaceName == NethubInterface.NETHUB_NAME) {
+            		_hostInterface = new NethubInterface();
+            		_hostInterface.RegisterReceiveCallback(OnHostPacketReceived);
+            		Console.WriteLine("** connected to nethub");
```
Would prefer not to have WriteLines in the code (use Log.Write where applicable.)  See other instances in this and other files as well.

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+		// this is the name of the NetHub network-device in the configuration dialog resp. the configuration file
+    	public const string NETHUB_NAME = "[[ Dodo-Nethub ]]";
+    	
+		private TcpClient client = null;
+		private NetworkStream stream = null;
+		
+		private ulong localAddress;
+		private ulong broadcastAddress;
+		
+		private ReceivePacketDelegate receiver = null;
+		private Thread receiverThread = null;
+		
+		private Byte[] sendBuffer = new byte[1026]; // this assumes that there is only ONE processing thread (which is the case for DarkStar!) 
```
Member variables should be prefixed with an underscore.

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        /// </summary>
+        /// <param name="callback"></param>
+        void IPacketInterface.RegisterReceiveCallback(ReceivePacketDelegate callback) {
+        	this.stopReceiverThread();
+        	this.receiver = callback;
+        	this.receiverThread = new Thread(this.PacketReceiver);
+        	this.receiverThread.Start();
```
No need for "this"

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        	Console.WriteLine("** Nethub interface shut down");
+        }
+        
+        private void stopReceiverThread() {
```
PascalCase please

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        	return (byte)value;
+        }
+        
+        private ulong readAddress(byte[] data, int at) {
```
PascalCase

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+        	
+        }
+        
+        private byte getByte(NetworkStream str) {
```
PascalCase

livingcomputermuseum commented on D/Ethernet/NethubInterface.cs:
```diff
@@ -0,0 +1,209 @@
+	        }
+        }
+        
+        private int getLenPrefix(NetworkStream str) {
```
PascalCase

livingcomputermuseum reviewed (changes requested):
See the comments in the code (mostly nitpicky, sorry).  Thanks for submitting this!

devhawala commented:
Hello,

no problem with pedantic code-reviews :-)

The changes in the new commit should incorporate your comments, hoping not to have overseen something.

For the 2 additional words for incoming packets, i removed the change in Darkstar, relying on the configuration option `ether.useDarkstarWorkaround` in Dodo until the real reason for the lost packet problem is found and fixed. Having thought about it, Dodo may have the issue.

Greetings, Hans

livingcomputermuseum reviewed (approved):
Looks good, thanks!
